plugin-enclave-gradle/src/main/kotlin/com/r3/conclave/plugin/enclave/gradle/GradleEnclavePlugin.kt

```diff
@@ -317,225 +232,206 @@ class GradleEnclavePlugin @Inject constructor(private val layout: ProjectLayout)
+                generateAppResourcesConfigTask,
+                createDummyKeyTask
+            )
+
+            val enclaveBundleJarTask = createTask<EnclaveBundleJar>("enclaveBundle${type}Jar", type) { task ->
+                // Based on the runtime config the relevant tasks will be executed. runtimeType.flatMap and
+                // runtimeType.map below automatically wire up the task dependencies so there's no need to manually
+                // call dependsOn.
+                task.bundleFile.set(runtimeType.flatMap {
+                    when (it) {
+                        GRAALVM -> signedGraalVMEnclaveFile
+                        GRAMINE -> gramineZipBundleTask.get().archiveFile
+                        null -> throw IllegalStateException()  // Keep the compiler happy
+                    }
+                })
+                task.fileName.set(runtimeType.map {
+                    val bundleName = when (it) {
+                        GRAALVM -> PluginUtils.GRAALVM_BUNDLE_NAME
+                        GRAMINE -> PluginUtils.GRAMINE_BUNDLE_NAME
+                        null -> throw IllegalStateException()  // Keep the compiler happy
+                    }
+                    "$typeLowerCase-$bundleName"
+                })
+                task.enclaveClassName.set(enclaveClassNameTask.enclaveClassName())
             }
 
-            val buildUnsignedEnclaveTask =
-                target.createTask<BuildUnsignedEnclave>("buildUnsignedEnclave$type") { task ->
-                    task.inputEnclave.set(buildUnsignedGraalEnclaveTask.outputEnclave)
-                    task.outputEnclave.set(task.inputEnclave.get())
-                }
+            project.artifacts.add(typeLowerCase, enclaveBundleJarTask.archiveFile)
+        }
+    }
 
-            val generateEnclaveConfigTask =
-                target.createTask<GenerateEnclaveConfig>("generateEnclaveConfig$type", type) { task ->
-                    task.productID.set(conclaveExtension.productID)
-                    task.revocationLevel.set(conclaveExtension.revocationLevel)
-                    task.maxHeapSize.set(conclaveExtension.maxHeapSize)
-                    task.maxStackSize.set(conclaveExtension.maxStackSize)
-                    task.tcsNum.set(conclaveExtension.maxThreads)
-                    task.outputConfigFile.set(enclaveDirectory.resolve("enclave.xml").toFile())
-                }
+    private fun createGraalVmTasks(
+        type: BuildType,
+        conclaveExtension: ConclaveExtension,
+        enclaveExtension: EnclaveExtension,
+        linuxExec: LinuxExec,
+        copyGraalVM: CopyGraalVM,
+        enclaveFatJarTask: EnclaveFatJar,
+        generateReflectionConfigTask: GenerateReflectionConfig,
+        generateAppResourcesConfigTask: GenerateAppResourcesConfig,
+        createDummyKeyTask: GenerateDummyMrsignerKey
+    ): Provider<RegularFile> {
+        fun buildFile(path: String): Provider<RegularFile> = file("${type.name.lowercase()}/$path")
```

Comment:
I don't understand this provider mechanism here. Why don't we just set the outputs with a File?